### PR TITLE
Predict ram

### LINKED ISSUE #48: function to estimate memory requirements
It would be helpfull if simulations/algorithms could estimate roughly how much memory they might need, such that one can reserve a suitable amount when running it on a computing cluster.

### PR DESCRIPTION
**Pull request by jangeiger**
This code predicts the RAM usage required to perform an algorithm in TeNPy (for example, DMRG).
Closes #48

The following features are implemented:
- [x] Implemented usage for MPS, MPO and environments
- [x] Implemented saving due to conservation laws
- [x] Extracted saving factor for particle number conservation in the Bose-Hubbard model

It is compatible with infinite and finite MPS system sizes and was benchmarked for the following cases:

**1) Finite 1D:**
- [x] Tested with finite 1D Ising model (Parity conservation)
- [x] Tested with finite 1D Bose-Hubbard model (Particle number conservation)

**2) Infinite 1D:**
- [x] Tested with infinite 1D Bose-Hubbard model (Particle number conservation)
- [x] Tested with infinite 1D Ising model (Parity conservation)

**3) 2D Lattice:**
- [x] Tested with 2D triangular Heisenberg model

Remark: All RAM tests were performed on the TUM cluster running on Linux. For finite size MPS, the length L was varied; for iMPS, the bond dimension was varied. The ram was extracted using psutil as well as tracemalloc.

### CODE REVIEW

File: tenpy/models/hubbard.py [hunk 1]
```diff
@@ -89,6 +89,30 @@ def __init__(self, model_params):
...
+        Returns
+        -------
+        factor : int
+            saving factor, due to conservation
+
+        Options
+        -------
+        .. cfg:configoptions :: Model
+
+            saving_factor :: None | int
```
Comment (Jakob-Unfried) [inline tenpy/models/hubbard.py:105]: Explicitly say that this is only used for RAM estimation via :func:`...Simulation.estimate_RAM`?

File: tenpy/models/model.py [hunk 2]
```diff
@@ -254,9 +254,33 @@ def update_time_parameter(self, new_time):
...
+        for mod in chinfo.mod:
+            if mod == 1:
+                savings *= 1/4 # this is what we found empirically
+            else:
+                savings *= 1/mod
+        return savings
 
 class NearestNeighborModel(Model):
-    r"""Base class for a model of nearest neigbor interactions w.r.t. the MPS index.
+    r"""Base class for a model of nearest neighbour interactions w.r.t. the MPS index.
```
Comment (Jakob-Unfried) [inline tenpy/models/model.py:283]: `neighbor` is the correct US spelling.

File: tests/test_predict_ram.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+"""Short test for vmem prediction."""
+# Copyright 2018-2023 TeNPy Developers, GNU GPLv3
```
Comment (Jakob-Unfried) [inline tests/test_predict_ram.py:2]: 2023

**Review by Jakob-Unfried — COMMENTED** [review]
Very nice!
I did not go through the internal logic in detail, it is only heuristics anyway and if it matches in the tests that you conducted, that is good enough for me.

Left some minor comments, let me know if anything is unclear

**Comment by Jakob-Unfried** [comment]
@jangeiger anything missing or can we merge?

**Comment by jangeiger** [comment]
If the code looks good to you, we can merge.

Apart from your comment on the config option, I have implemented everything else. I agree that the saving factor is not a physical parameter. If you wish, we can remove this option. However, for a more accurate prediction, one would probably want to automatically extract it via a small-scale DMRG simulation anyway.

**Comment by Jakob-Unfried** [comment]
While reading this again, I found that `estimate_RAM_saving_factor` looks buggy. Should theta not be an `else` clause that deals with `mod != 1`?

I have two main worries regarding the `saving_factor` as a config option:

- Very few people will know the models and charges intimately enough to give good values. I certainly would not trust myself to give a good guess without running numerical experiments. The whole point of this feature is to move away from peoples intuition and to have some hard-coded heuristics.

- The option will appear in the docs of all models, e.g. [here](https://tenpy.readthedocs.io/en/latest/reference/tenpy.models.toric_code.ToricCode.html#cfg-config-ToricCode). That means that if you give model parameters somewhere, you might encounter this option, and in most cases users should just leave it alone.

I suggest to simply remove the option and rely on the heuristics that you found.

If there is some way to get a better heuristic, we should merge it into tenpy if it concerns one of the models we implement, or the user can simply overwrite `estimate_RAM_saving_factor` for their particular model.

**Comment by Jakob-Unfried** [comment]
Other than that, we are ready to merge =)

**Comment by jangeiger** [comment]
I agree this option is confusing - I have removed it 👍

**Pull request merged**